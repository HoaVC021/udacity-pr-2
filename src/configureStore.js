import { applyMiddleware, createStore } from 'redux';
import { thunk } from 'redux-thunk';

import loggerMiddleware from './middlewares/logger';
import reducer from './reducers';

export default function configureStore(preloadedState) {
  const middlewares = [loggerMiddleware, thunk];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(reducer, preloadedState, middlewareEnhancer);

  return store;
}
