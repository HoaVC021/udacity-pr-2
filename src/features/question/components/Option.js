import React, { useCallback } from 'react';
import { Box, Button, Typography } from '@mui/material';

const Option = ({ optionName, answered, selectedOption, option, totalVotes, onVote }) => {

  const handleClick = useCallback(() => {
    onVote(optionName);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Box
      sx={{
        borderRadius: (theme) => `${theme.spacing(1)}`,
        border: (theme) => `1px solid ${optionName === selectedOption && answered ? theme.palette.success.main : theme.palette.grey[300]}`,
      }}
    >
      <Box sx={{ my: 2, width: '100%' }}>
        <Typography variant='body1' textAlign='center'>{option.text}</Typography>
        {answered && (
          <>
            <Typography variant='body2' textAlign='center'>
              {option.votes.length} votes
            </Typography>
            <Typography variant='body2' textAlign='center'>
              {((option.votes.length / totalVotes) * 100).toFixed(2)}% votes
            </Typography>
          </>
        )}
      </Box>
      <Button onClick={handleClick} fullWidth variant='contained' color='primary' size='large' disabled={answered}>
        {answered ? 'Voted' : 'Click'}
      </Button>
    </Box>
  );
};

export default Option;
