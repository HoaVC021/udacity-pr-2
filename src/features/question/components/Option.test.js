import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Option from './Option'; // Adjust the import path as necessary
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);
describe('Option Component', () => {
    const mockOnVote = jest.fn();

    const setup = (props) => {
        return render(<Option {...props} onVote={mockOnVote} />);
    };

    beforeEach(() => {
        jest.clearAllMocks(); // Clear previous mock calls
    });

    test('renders correctly with unanswered option', () => {
        const props = {
            optionName: 'optionOne',
            answered: false,
            selectedOption: '',
            option: { text: 'Option 1', votes: [] },
            totalVotes: 0,
        };

        setup(props);

        expect(screen.getByText(/option 1/i)).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /click/i })).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /click/i })).toBeEnabled();
    });

    test('renders correctly with answered option', () => {
        const props = {
            optionName: 'optionOne',
            answered: true,
            selectedOption: 'optionOne',
            option: { text: 'Option 1', votes: ['user1', 'user2'] },
            totalVotes: 2,
        };

        setup(props);

        expect(screen.getByText(/option 1/i)).toBeInTheDocument();
        expect(screen.getByText(/2 votes/i)).toBeInTheDocument();
        expect(screen.getByText(/100.00% votes/i)).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /voted/i })).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /voted/i })).toBeDisabled();
    });

    test('handles click event correctly', () => {
        const props = {
            optionName: 'optionOne',
            answered: false,
            selectedOption: '',
            option: { text: 'Option 1', votes: [] },
            totalVotes: 0,
        };

        setup(props);

        fireEvent.click(screen.getByRole('button', { name: /click/i }));

        expect(mockOnVote).toHaveBeenCalledWith('optionOne');
    });

    it("should match snapshot", () => {
        const mockOnVote = jest.fn();
        const { asFragment } = render(
            <Option
                optionName="optionOne"
                answered={false}
                selectedOption=""
                option={{ text: 'Option 1', votes: [] }}
                totalVotes={0}
                onVote={mockOnVote}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
