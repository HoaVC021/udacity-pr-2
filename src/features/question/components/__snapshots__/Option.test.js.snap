// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-abjdan"
  >
    <div
      class="MuiBox-root css-uawqwr"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1699v82-MuiTypography-root"
      >
        Option 1
      </p>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-fullWidth css-w9wzph-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Click
    </button>
  </div>
</DocumentFragment>
`;
